File: AGRA_0.5/Hw04/Ejemplos/gabow2.cpp
/*
Algoritmo de Gabow (SCC)
Autor: Carlos Ramirez
Fecha: Marzo 23 de 2020

 */

#include <vector>
#include <stack>
#include <iostream>

using namespace std;

int n, numSCC, t;
vector<vector<int> > adj(50000);
int visitado[50000];
int sccInd[50000];
stack<int> pilaS, pilaP;
vector<vector<int> > sccNodos;

void gabowAux(int);

void gabow(){
  int i;

  for(i = 0; i < n; i++)
    sccInd[i] = visitado[i] = -1;

  for(i = 0; i < n; i++)
    if(visitado[i] == -1)
      gabowAux(i);
}

void gabowAux(int v){
  int w;
  visitado[v] = ++t;
  pilaS.push(v);
  pilaP.push(v);
  
  for(int i = 0; i < adj[v].size(); i++){
    w = adj[v][i];
    if(visitado[w] == -1)
      gabowAux(w);
    else if(sccInd[w] == -1){
      while(visitado[pilaP.top()] > visitado[w])
	pilaP.pop();
    }
  }

  if(v == pilaP.top()){
    numSCC++;
    sccNodos.push_back(vector<int>());
    cout << "SCC con indice " << numSCC << ": ";
    while(pilaS.top() != v){
      cout << pilaS.top() << " ";
      sccInd[pilaS.top()] = numSCC - 1;
      sccNodos[numSCC - 1].push_back(pilaS.top());
      pilaS.pop();
    }

    cout << pilaS.top() << endl;
    sccInd[pilaS.top()] = numSCC - 1;
    sccNodos[numSCC - 1].push_back(pilaS.top());
    pilaS.pop();
    pilaP.pop();
  }
}

int main(){
  int m, i, aux1, aux2;

  cin >> n >> m;

  for(i = 0; i < m; i++){
    cin >> aux1 >> aux2;
    adj[aux1].push_back(aux2);
  }

  gabow();
  
  return 0;
}

File: AGRA/Hw04/Ejemplos/gabow2.py
from sys import stdin

MAX = 10000
adj = [[] for i in range(MAX)]
visitado = [0 for i in range(MAX)]
sccInd = [-1 for i in range(MAX)]
n, t, numSCC = int(), 0, 0
sccNodos, pilaS, pilaP = [], [], []

def gabow():
    for i in range(n):
        sccInd[i], visitado[i] = -1, -1

    for i in range(n):
        if visitado[i] == -1:
            gabowAux(i)

def gabowAux(v):
    global t, numSCC
    t += 1
    visitado[v] = t
    pilaS.append(v)
    pilaP.append(v)

    for i in range(len(adj[v])):
        w = adj[v][i];
        if visitado[w] == -1:
            gabowAux(w)
        elif sccInd[w] == -1:
            while visitado[pilaP[-1]] > visitado[w]:
                pilaP.pop()

    if v == pilaP[-1]:
        numSCC += 1
        sccNodos.append([])
        print("SCC con indice %d: " % numSCC, end = '')
        while pilaS[-1] != v:
            a = pilaS.pop()
            print("%d " % a, end = '')
            sccInd[a] = numSCC - 1
            sccNodos[numSCC - 1].append(a)

        a = pilaS.pop()
        print("%d " % a)
        sccInd[a] = numSCC - 1
        sccNodos[numSCC - 1].append(a)
        pilaP.pop()

def main():
    global n
    n, m = list(map(int, stdin.readline().split()))

    for i in range(m):
        u, v = list(map(int, stdin.readline().split()))
        adj[u].append(v)

    print("Grafo")
    for i in range(n):
        print("Nodo %d:" % i)
        for j in range(len(adj[i])):
            print("%d" % adj[i][j], end = ' ')
        print("")

    print("Componentes Fuertemente Conexos:")
    gabow()

main()

    
    
